zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java

```diff
@@ -81,7 +87,32 @@ public abstract class X509Util implements Closeable, AutoCloseable {
         }
     }
 
-    public static final String DEFAULT_PROTOCOL = "TLSv1.2";
+    public static final String DEFAULT_PROTOCOL = defaultTlsProtocol();
+
+    /**
+     * Return TLSv1.3 or TLSv1.2 depending on Java runtime version being used.
+     * TLSv1.3 was first introduced in JDK11 and back-ported to OpenJDK 8u272.
+     */
+    private static String defaultTlsProtocol() {
+        String defaultProtocol = "TLSv1.2";
+        List<String> supported = new ArrayList<>();
+        try {
+            supported = Arrays.asList(SSLContext.getDefault().getSupportedSSLParameters().getProtocols());
+            if (supported.contains("TLSv1.3")) {
+                defaultProtocol = "TLSv1.3";
+            }
+        } catch (NoSuchAlgorithmException e) {
+            // Ignore.
+        }
+        LOG.info("Default TLS protocol is {}, supported TLS protocols are {}", defaultProtocol, supported);
+        return defaultProtocol;
+    }
+
+    // ChaCha20 was introduced in OpenJDK 11.0.15 and it is not supported by JDK8.
+    private static String[] getTLSv13Ciphers() {
+        return new String[]{"TLS_AES_256_GCM_SHA384", "TLS_AES_128_GCM_SHA256", "TLS_CHACHA20_POLY1305_SHA256"};
+    }
```

Comment:
Exactly, this whole problem _began_ from hardcoded ciphers in ZK, https://issues.apache.org/jira/browse/ZOOKEEPER-4415?focusedCommentId=17497994&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17497994 but did not dare to change that - so in a way building on top what existed.

Reply:
Since the cipher list is hardcoded, I needed to extend it with TLSv1.3 ciphers. My approach was to list all TLSv1.3 ciphers that JDK would normally enable by default. So, I created the list by checking what `SSLServerSocketFactory.getDefault().getDefaultCipherSuites()` returns on recent JDK, and out from that list I picked the ones that are for TLSv1.3. This also happens to match with list recommended here https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility.

There are more TLSv1.3 ciphers in JDK, but Java does not enable them by default  https://docs.oracle.com/en/java/javase/17/docs/specs/security/standard-names.html#jsse-cipher-suite-names (see rows where column "Introduced in" says TLSv1.3).

Reply:
> We added the 2 optimized list of cipher suites for performance reasons. GCM ciphers are better under JDK8, while CBC ciphers are better with JDK9+. We support all of them, just building a list for cipher preference.

Given that TLS is off by default, I suspect people who turn it on are doing so for security considerations. Performance is a deciding factor, but I think it's trying to read the mind of users by making a pre-made set, and further deciding what is "better" for a given JDK. GCM and CBC are very different in terms of the security features.

I would leave the optimizations to the user's JDK distro and just stick with the defaults, rather than trying to override it with hard-coded lists. It'd be better to just write a simple doc that makes suggestions for users, or point them to an external resource that stays up to date with the latest in security news, if that's a concern, rather than hard-code your own lists.